### PR TITLE
Add utility riak_array

### PR DESCRIPTION
**Pull request by hazen**
- Also run integration tests only when command-line argument is specified
- Used as a key datatype in upcoming connection pool implementation

### CODE REVIEW

File: src/include/riak_array.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    riak_config *cfg;
+    riak_int64_t length;
+    riak_int64_t capacity;
+    void       **array;
+} riak_array;
+
+/**
+ * @brief Create yet another generic C data structure
+ * @param cfg Riak Configuration
+ * @param size Initial size of the array
```
Comment (bookshelfdave) [inline src/include/riak_array.h]: size => capacity?

File: src/riak_array.c [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+riak_array_new(riak_config *cfg,
+               riak_size_t  size) {
+    riak_array *ra = riak_config_clean_allocate(cfg, sizeof(riak_array));
+    if (ra != NULL) {
+        ra->cfg      = cfg;
+        ra->length   = 0;
+        ra->capacity = size;
+        ra->array    = riak_config_clean_allocate(cfg, sizeof(void*)*size);
+        if (ra->array == NULL) {
+            riak_free(cfg, &ra);
```
Comment (bookshelfdave) [inline src/riak_array.c]: should debug logging be added here?

**Comment by bookshelfdave** [comment]
going to Valgrind this in BOS

File: src/include/riak_array.h (hunk 1 above)
Comment (hazen) [inline src/include/riak_array.h]: Yes, this is capacity. I'll change the name.

File: src/riak_array.c (hunk 2 above)
Comment (hazen) [inline src/riak_array.c]: Typically I just return an error to the caller when memory fails.  I can add logging if you think it would be clearer.

Comment (bookshelfdave) [inline src/riak_array.c]: how would you feel about making `riak_array_new` take an additional `riak_foo **foo_target` parameter, and change its return type to riak_error, similar to how other _new() functions work in the RCC?

Comment (hazen) [inline src/riak_array.c]: As you can probably tell I vacillate quite a bit on this one.  I really should be consistent.  `riak_error` it is.

**Comment by hazen** [comment]
My test with valgrind on `riak_c_cunit` doing ONLY the `riak_array` suite shows leaks only in libevent:

```
==18105==
==18105== HEAP SUMMARY:
==18105==     in use at exit: 216 bytes in 6 blocks
==18105==   total heap usage: 96 allocs, 90 frees, 17,916 bytes allocated
==18105==
==18105== 32 bytes in 1 blocks are still reachable in loss record 1 of 6
==18105==    at 0x4C2C857: malloc (vg_replace_malloc.c:291)
==18105==    by 0x54A6E79: debug_lock_alloc (evthread.c:161)
==18105==    by 0x54A6A42: event_global_setup_locks_ (event.c:2882)
==18105==    by 0x409DCE: main (registry.c:56)
==18105==
==18105== 32 bytes in 1 blocks are still reachable in loss record 2 of 6
==18105==    at 0x4C2C857: malloc (vg_replace_malloc.c:291)
==18105==    by 0x54A6E79: debug_lock_alloc (evthread.c:161)
==18105==    by 0x54B87A3: evsig_global_setup_locks_ (signal.c:442)
==18105==    by 0x54A6A55: event_global_setup_locks_ (event.c:2884)
==18105==    by 0x409DCE: main (registry.c:56)
==18105==
==18105== 32 bytes in 1 blocks are still reachable in loss record 3 of 6
==18105==    at 0x4C2C857: malloc (vg_replace_malloc.c:291)
==18105==    by 0x54A6E79: debug_lock_alloc (evthread.c:161)
==18105==    by 0x54B6953: evutil_secure_rng_global_setup_locks_ (evutil_rand.c:121)
==18105==    by 0x54A6A67: event_global_setup_locks_ (event.c:2886)
==18105==    by 0x409DCE: main (registry.c:56)
==18105==
==18105== 40 bytes in 1 blocks are still reachable in loss record 4 of 6
==18105==    at 0x4C2C857: malloc (vg_replace_malloc.c:291)
==18105==    by 0x5291DB9: evthread_posix_lock_alloc (evthread_pthread.c:46)
==18105==    by 0x54A7527: evthread_setup_global_lock_ (evthread.c:358)
==18105==    by 0x54A6A42: event_global_setup_locks_ (event.c:2882)
==18105==    by 0x5291F5F: evthread_use_pthreads (evthread_pthread.c:185)
==18105==    by 0x409DD3: main (registry.c:57)
==18105==
==18105== 40 bytes in 1 blocks are still reachable in loss record 5 of 6
==18105==    at 0x4C2C857: malloc (vg_replace_malloc.c:291)
==18105==    by 0x5291DB9: evthread_posix_lock_alloc (evthread_pthread.c:46)
==18105==    by 0x54A7527: evthread_setup_global_lock_ (evthread.c:358)
==18105==    by 0x54B87A3: evsig_global_setup_locks_ (signal.c:442)
==18105==    by 0x54A6A55: event_global_setup_locks_ (event.c:2884)
==18105==    by 0x5291F5F: evthread_use_pthreads (evthread_pthread.c:185)
==18105==    by 0x409DD3: main (registry.c:57)
==18105==
==18105== 40 bytes in 1 blocks are still reachable in loss record 6 of 6
==18105==    at 0x4C2C857: malloc (vg_replace_malloc.c:291)
==18105==    by 0x5291DB9: evthread_posix_lock_alloc (evthread_pthread.c:46)
==18105==    by 0x54A7527: evthread_setup_global_lock_ (evthread.c:358)
==18105==    by 0x54B6953: evutil_secure_rng_global_setup_locks_ (evutil_rand.c:121)
==18105==    by 0x54A6A67: event_global_setup_locks_ (event.c:2886)
==18105==    by 0x5291F5F: evthread_use_pthreads (evthread_pthread.c:185)
==18105==    by 0x409DD3: main (registry.c:57)
==18105==
==18105== LEAK SUMMARY:
==18105==    definitely lost: 0 bytes in 0 blocks
==18105==    indirectly lost: 0 bytes in 0 blocks
==18105==      possibly lost: 0 bytes in 0 blocks
==18105==    still reachable: 216 bytes in 6 blocks
==18105==         suppressed: 0 bytes in 0 blocks
==18105==
==18105== For counts of detected and suppressed errors, rerun with: -v
==18105== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 2 from 2)
```

I've looked in the source and the documentation, but I don't see any obvioius "clean up global locks" function.

If I comment out ALL libevent from the unit test:

```
==18536==
==18536== HEAP SUMMARY:
==18536==     in use at exit: 0 bytes in 0 blocks
==18536==   total heap usage: 90 allocs, 90 frees, 17,700 bytes allocated
==18536==
==18536== All heap blocks were freed -- no leaks are possible
==18536==
==18536== For counts of detected and suppressed errors, rerun with: -v
==18536== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 2 from 2)
```

File: src/include/riak_array.h [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+/**
+ * @brief Free any memory claimed by the array itself
+ * NOTE: Elements must be freed separately
+ * @param ra Riak Array
+ */
+void
+riak_array_free(riak_array **ra);
+
+/**
+ * @brief Access the size of the underlying array
```
Comment (bookshelfdave) [inline src/include/riak_array.h]: Access the "length", we already have length and capacity.

File: src/include/riak_array.h [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+ * @param ra Riak Array
+ * @returns The number of elements in the Riak Array
+ */
+riak_size_t
+riak_array_length(riak_array *ra);
+
+/**
+ * @brief Add an element to an existing array (shifting other elements forward)
+ * @param ra Riak Array
+ * @param pos Position within array to set value
```
Comment (bookshelfdave) [inline src/include/riak_array.h]: indexed starting from 0?

File: src/include/riak_array.h [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+ * @param pos Position at which to remove an element
+ * @returns The selected element
+ */
+void*
+riak_array_remove(riak_array *ra,
+                  riak_size_t pos);
+
+/**
+ * @brief Add an element to an existing array (at the end)
+ * @param ra Riak Array
```
Comment (bookshelfdave) [inline src/include/riak_array.h:85]: can you doc what happens if the length > capacity?

File: src/include/riak_array.h [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+ * @returns The selected element
+ */
+void*
+riak_array_pop(riak_array *ra);
+
+/**
+ * @brief Access an element from a Riak Array
+ * @param ra Riak Array
+ * @param pos Position at which to retrieve an element
+ * @returns The selected element
```
Comment (bookshelfdave) [inline src/include/riak_array.h]: what happens if an invalid pos is specified?

File: src/riak_array.c [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    ra->array[pos] = elem;
+
+    return ERIAK_OK;
+}
+
+void*
+riak_array_remove(riak_array *ra,
+                  riak_size_t pos) {
+    if ((riak_int64_t)pos > ra->length-1) {
+        return NULL;
```
Comment (bookshelfdave) [inline src/riak_array.c]: NULL is signaling an error, should this fn return `riak_error` instead?

File: src/riak_array.c [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    return ERIAK_OK;
+}
+
+void*
+riak_array_pop(riak_array *ra) {
+    if (ra->length > 0) {
+        ra->length--;
+        return ra->array[ra->length];
+    } else {
+        return NULL;
```
Comment (bookshelfdave) [inline src/riak_array.c]: same thing here, NULL is an error.

Comment (bookshelfdave) [inline src/riak_array.c]: ERIAK_OUT_OF_RANGE?

File: test/cunit/test_array.c [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+}
+
+void
+test_array_push() {
+    riak_config *cfg;
+    riak_error err = riak_config_new_default(&cfg);
+    CU_ASSERT_EQUAL_FATAL(err,ERIAK_OK)
+
+    riak_array *ra;
+    err = riak_array_new(cfg, &ra, 10);
```
Comment (bookshelfdave) [inline test/cunit/test_array.c:66]: the array is created with a cap of 10, but you are inserting 100 elements. Should you note that this is testing the resize functionality?

File: test/cunit/test_array.c [hunk 10]
```diff
@@ -0,0 +1,283 @@
...
+    CU_ASSERT_EQUAL_FATAL(err,ERIAK_OK)
+
+    for(riak_uint64_t i = 1; i <= 100; i++) {
+    err = riak_array_insert(ra, 0, (void*)i);
```
Comment (bookshelfdave) [inline test/cunit/test_array.c]: tab me!

File: test/cunit/test_array.c [hunk 11]
```diff
@@ -0,0 +1,283 @@
...
+        CU_ASSERT_EQUAL_FATAL(err,ERIAK_OK)
+    }
+    // Overwrite 100 values with i-1
+    for(riak_uint64_t i = 0; i < 100; i++) {
+        riak_uint64_t old_value = (riak_uint64_t)riak_array_get(ra, i);
+    err = riak_array_set(ra, i, (void*)i);
```
Comment (bookshelfdave) [inline test/cunit/test_array.c]: tab me!

File: src/riak_array.c (hunk 7 above)
Comment (bookshelfdave) [inline src/riak_array.c]: I don't feel strongly either way, I see the tests below check the return value etc.

**Comment by bookshelfdave** [comment]
looks good

File: src/riak_array.c (hunk 7 above)
Comment (hazen) [inline src/riak_array.c]: I've been writing code against this and think that `riak_error` is a better approach.  Rewriting heading your way soon

File: test/cunit/test_array.c (hunk 9 above)
Comment (hazen) [inline test/cunit/test_array.c:66]: I would have guessed that pushing and inserting would grow the array.  I'm happy to add that to the test name, however

Comment (bookshelfdave) [inline test/cunit/test_array.c:66]: just a comment will do :-)

**Comment by bookshelfdave** [comment]
bah, feel free to ignore my doc return type comments. Like you say, they'll probably just get out of sync with the code.

```
============================================================================
Testsuite summary for riak_c_client 0.5
============================================================================
# TOTAL: 1
# PASS:  1
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
```

**Comment by bookshelfdave** [comment]
:+1:

**Pull request merged**